dataset_example.py

```diff
@@ -0,0 +1,161 @@
+import typing
+from abc import ABC
+
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+
+from flytekit import kwtypes, task, workflow
+from flytekit.types.structured.structured_dataset import (
+    FLYTE_DATASET_TRANSFORMER,
+    DatasetDecodingHandler,
+    DatasetEncodingHandler,
+    DatasetFormat,
+    StructuredDataset,
+)
+
+PANDAS_PATH = "/tmp/pandas.pq"
+NUMPY_PATH = "/tmp/numpy.pq"
+BQ_PATH = "bq://photo-313016:flyte.new_table3"
+
+# https://github.com/flyteorg/flyte/issues/523
+# We should also support map and list in schema column
+column_type = kwtypes(w=typing.Dict[str, typing.Dict[str, int]], x=typing.List[typing.List[int]], y=int, z=str)
+
+
+@task
+def t0() -> pd.DataFrame:
+    return pd.DataFrame({"Name": ["Tom", "Joseph"], "Age": [20, 22]})
+
+
+@task
+def t1(dataframe: pd.DataFrame) -> StructuredDataset[column_type]:
+    # S3 (parquet) -> Pandas -> S3 (parquet) default behaviour
+    return dataframe
+
+
+@task
+def t1a(dataframe: pd.DataFrame) -> StructuredDataset[column_type]:
+    # S3 (parquet) -> Pandas -> S3 (parquet) default behaviour
+    return StructuredDataset(dataframe=dataframe, remote_path=PANDAS_PATH)
+
+
+@task
+def t2(dataframe: pd.DataFrame) -> pd.DataFrame:
+    # S3 (parquet) -> Pandas -> S3 (parquet)
+    return dataframe
+
+
+@task
+def t3(dataset: StructuredDataset[column_type]) -> StructuredDataset[column_type]:
+    # s3 (parquet) -> pandas -> s3 (parquet)
+    print("Pandas dataframe")
+    print(dataset.open_as(pd.DataFrame))
+    return dataset
```

Comment:
let's add a comment saying that this should not upload anything, since we're returning just the input object.